### PR TITLE
Make `ignoreLink` accessible for plugins

### PR DESCRIPTION
**Pull request by hirasso**
- move all checks out of `linkClickHandler` and in a dedicated method `ignoreLink`
- move the `el.origin` check from the option inside this method

Makes it possible for plugins to call `swup.ignoreLink(el)`

Related: #551

### CODE REVIEW

**Comment by hirasso** [comment]
> 1 — Let's rename it to sound like a conditional: shouldIgnoreLink maybe. Right now it sounds like calling it will make swup ignore a specific link.

Good call, let's prefix the function with `should`!

Regarding your suggestion: You definitely have a broader oversight over swups plugin ecosystem. I'm not able to fully think it through without seeing the actual implementation – but your reasoning sounds plausible, would you like to give it a shot?

```js 
shouldIgnoreVisit({ href, el }) {
    // ...
}
```
How would you imagine that this would be called? Intuitively would have gone for a required `href` and an optional `el` parameter, something like this:
```js 
shouldIgnoreVisit(href, el = null) {
    // ...
}
```

**Comment by daun** [comment]
Sure, I'll give it a try.

The reasoning behind the destructured object argument was to be able to add parameters later if required. But I like the required href param, so let's do both. That way we can call it with just a string as well.

```js
shouldIgnoreVisit(href, { el } = {}) {

}
```

**Comment by hirasso** [comment]
Nice! Looking forward.

**Comment by daun** [comment]
I've pushed a few changes.

- Rename `shouldIgnoreVisit` method
- Rename `options.ignoreLink` option to `ignoreVisit` to make its purpose clearer. Need to change docs as well, and your PR on the preload plugin.
- Pass around the `el` as optional parameter

Now we can check for forms as well:

```js
const form = event.delegateTarget;
const href = form.action;
if (this.swup.shouldIgnoreVisit(href, { el: form }) {
  //
}
```

I've used a `URL` object here for the origin check, I'd like to do a refactor to use that exclusively but that's another PR :)

**Comment by daun** [comment]
What's great about this PR is any plugin can now extend and overwrite this method on the swup instance.

**Comment by hirasso** [comment]
Looking good! I'll test it tomorrow and report back.

File: src/index.js
Comment (hirasso) [inline src/index.js:33]: Shouldn't we make the second (options) parameter optional here as well?

```js
ignoreVisit: (href, { el } = {}) => el?.closest('[data-no-swup]'),
```

Comment (daun) [reply src/index.js:33]: Sure, why not. Feel free to push :)

**Comment by hirasso** [comment]
Good to go from my side :) I'll move over to preload plugin then to adjust it the new method name.

**Comment by daun** [comment]
Great! Can you merge? I'm away from the keyboard until the end of next week.

**Comment by hirasso** [comment]
These duration tests are always making trouble in the ci... 🤖🐌

**Comment by hirasso** [comment]
> Great! Can you merge? I'm away from the keyboard until the end of next week.

I can't merge without your approval. But no rush, the tests are failing again anyways :)

**Review by daun — APPROVED** [review]

**Comment by hirasso** [comment]
So it seems to me like the tests sometimes  fail because the whole duration is *sometimes* the duration plus the delay and sometimes only the duration. Could this be the cause somehow? I find it quite hard to debug the tests...

**Comment by hirasso** [comment]
Did some more investigation. The implementation of the transition animation seems off to me. I created a manual test to understand it's current behavior: It starts the animation with a duration of 400ms after a delay of 300ms: 
```css
animation: test-animation 400ms 300ms;
```
The animation fades the content out and in again:
```css
@keyframes test-animation {
  0%, 100% {
    transform: translateY(0);
    opacity: 1;
  }
  50% {
    transform: translateY(10px);
    opacity: 0;
  }
}
```
This results in the container being faded out and immediately in again, and only after the animation finished it replaces the content in the container. Not what we want, right? :)

I would presume the way we want keyframe animations to be used with swup would rather look something like this:
```css
@keyframes test-animation {
  from {
    transform: translateY(0);
    opacity: 1;
  }
  to {
    transform: translateY(10px);
    opacity: 0;
  }
}

html.is-changing .transition-main {
  animation: test-animation 300ms;
}
html.is-rendering .transition-main {
  animation-direction: reverse;
}
```
For the test case we could add the `animation-delay` back to the animation, like this:
```css
html.is-changing .transition-main {
  animation: test-animation 300ms 200ms; /* should result in 500ms total duration */
}
```

**Comment by hirasso** [comment]
Tests are passing now, I'll merge and we can tackle the tests in another PR should we need to.

**Comment by daun** [comment]
👽

**Pull request merged**
